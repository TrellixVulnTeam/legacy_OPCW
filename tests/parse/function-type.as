# Functions are first-class objects and are a valid type for a
# slot declaration.
let static_fn: delegate() = name;

# A more complex function type.
let more_fn: delegate(int, bool) -> int;

# A function type that takes a function and
# returns a function that returns an int.
let fn_fn: delegate(delegate(int) -> int) -> delegate(int) -> int;

# A named function that fulfils the above type.
let fn_fn_named(fn: delegate(int) -> int): (delegate(int) -> int) -> {
    fn;
}
