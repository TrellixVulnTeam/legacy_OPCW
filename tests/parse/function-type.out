ModuleDecl <?> _
  Nodes <?> 
    SlotDecl <?> static_fn
      Delegate <?>
        Parameters <?> 
      Ident <?> name
    SlotDecl <?> more_fn
      Delegate <?>
        Ident <?> int
        Parameters <?> 
          FuncParam <?>
            Ident <?> int
          FuncParam <?>
            Ident <?> bool
    SlotDecl <?> fn_fn
      Delegate <?>
        Delegate <?>
          Ident <?> int
          Parameters <?> 
            FuncParam <?>
              Ident <?> int
        Parameters <?> 
          FuncParam <?>
            Delegate <?>
              Ident <?> int
              Parameters <?> 
                FuncParam <?>
                  Ident <?> int
    FuncDecl <?> fn_fn_named
      Delegate <?>
        Ident <?> int
        Parameters <?> 
          FuncParam <?>
            Ident <?> int
      Type Parameters <?> 
      Parameters <?> 
        FuncParam <?> fn
          Delegate <?>
            Ident <?> int
            Parameters <?> 
              FuncParam <?>
                Ident <?> int
      Block <?>
        Nodes <?> 
          Ident <?> fn
